proposals/0000-no-implicit-binds.rst

```diff
@@ -0,0 +1,446 @@
+``-XNoImplicitForAll`` alone would *not* cause ``t`` in the signature to be bound from in the above; it would be unbound causing and error.
+But, as a newly freed variable, it is now ready to be captured by whatever is proposed in `#270`_.
+I think this is a good separation of concerns.
+
+Proposed Change Specification
+-----------------------------
+
+Create ``-XImplicitForAll`` to allow automatically capturing free variables in an outer ``forall`` as is always done today.
+It is on by default for backwards compatibility.
+When using ``-XNoImplicitForAll``, all variables in regular signatures, instances, and data declerations must be explicitly bound.
+
+Create ``-XPatternSignatureBinds`` to allow the implicit binding of free variables in pattern signatures.
+It is also on by default for backwards compatibility.
+When using ``-XNoPatternSignatureBinds``, all variables in pattern signatures must be explicitly bound.
+
+Examples
+--------
+
+Basic examples
+~~~~~~~~~~~~~~
+
+#. ::
+
+     {-# LANGUAGE NoImplicitForAll #-}
+     -- {-# LANGUAGE NoPatternSignatureBinds #-} -- Does not matter whether enabled or disabled
+
+     f :: t -> ... -- error: `t` is not bound
+     f x = ...
+
+   This could be rewritten as::
+
+     {-# LANGUAGE NoImplicitForAll #-}
+     -- {-# LANGUAGE PatternSignatureBinds #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE ExplicitForAll #-}
+
+     f :: forall t. t -> ...
+     f x = ...
+
+#. ::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+
+     f (x :: t) = ... -- error: `t` is not bound
+
+   This could be rewritten as either::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+     {-# LANGUAGE ScopedTypeVariables #-}
+
+     f :: forall t. ...
+     f (x :: t) = ... -- OK
+
+   or, if `#238`_ is accepted::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+     -- {-# LANGUAGE ScopedTypeVariables #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE ExplicitForAll #-}
+     {-# LANGUAGE TypeApplications #-}
+     {-# LANGUAGE TypeAbstractions #-} -- from #238
+
+     f :: forall t0. ...
+     f @t (x :: t) = ... -- OK
+
+#. ::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+     {-# LANGUAGE ExistentialQuantification #-}
+
+     data Some where
+       MkSome :: forall t. t -> Some
+
+     f (MkSome (x :: t)) = ... -- error: `t` is not bound
+
+   This could be rewritten as::
+
+     -- {-# LANGUAGE NoImplicitForAll #-} -- Does not matter whether enabled or disabled
+     {-# LANGUAGE NoPatternSignatureBinds #-}
+     {-# LANGUAGE ExistentialQuantification #-}
+     {-# LANGUAGE TypeApplications #-}
```

Comment:
Don't you need the type-application-in-pattern extension from a recent proposal rather than `-XTypeApplications`?

Reply:
https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0126-type-applications-in-patterns.rst#proposed-change-specification there's isn't one, actually.

Reply:
I linked it for good measure